pkg/controller/miganalytic/volume_adjustment.go

```diff
@@ -0,0 +1,249 @@
+			matched := percentageMatcher.FindStringSubmatch(cols[4])
+			if len(matched) > 1 {
+				pv.UsagePercentage, err = strconv.ParseInt(matched[1], 10, 64)
+				pv.IsError = (err != nil)
+			}
+			return
+		}
+	}
+	for _, line := range stdErrLines {
+		if lineMatcher.MatchString(line) {
+			pv.IsError = true
+			return
+		}
+	}
+	return
+}
+
+// PrepareDFCommand given a list of volumes, creates a bulk df command for all volumes
+func (cmd *DFCommand) PrepareDFCommand(pvcs []MigAnalyticPersistentVolumeDetails) []string {
+	command := []string{
+		"/bin/bash",
+		"-c",
+	}
+	volPaths := []string{}
+	for _, pvc := range pvcs {
+		volPaths = append(volPaths,
+			fmt.Sprintf(
+				DFDiskPath,
+				cmd.BaseLocation,
+				pvc.PodUID,
+				pvc.VolumeName))
+	}
+	return append(command, fmt.Sprintf("df -B%s %s", cmd.BlockSize, strings.Join(volPaths, " ")))
+}
+
+// findOriginalPVDataMatchingDFOutput given a df output for a pv and nested map of nodeName->[]pvc, finds ref to matching object in the map
+func (pva *PersistentVolumeAdjuster) findOriginalPVDataMatchingDFOutput(pvc DFOutput, pvcNodeMap map[string][]MigAnalyticPersistentVolumeDetails) *MigAnalyticPersistentVolumeDetails {
+	if _, exists := pvcNodeMap[pvc.Node]; exists {
+		for i := range pvcNodeMap[pvc.Node] {
+			pvDetailRef := &pvcNodeMap[pvc.Node][i]
+			if pvc.Name == pvDetailRef.Name && pvc.Namespace == pvDetailRef.Namespace {
+				return pvDetailRef
+			}
+		}
+	}
+	return nil
+}
+
+// generateWarningForErroredPVs given a list of dfoutputs, generates warnings for associated pvcs
+func (pva *PersistentVolumeAdjuster) generateWarningForErroredPVs(erroredPVs []*DFOutput) {
+	warningString := "Failed gathering extended PV usage information for PVs [%s]"
+	pvNames := ""
+	for _, dfOutput := range erroredPVs {
+		if dfOutput.IsError {
+			fmt.Sprintf(
+				pvNames,
+				dfOutput.Name)
+		}
+	}
+	pva.Owner.Status.Conditions.SetCondition(migapi.Condition{
+		Category: migapi.Warn,
+		Status: True,
+		Type: ExtendedPVAnalysisFailed,
+		Reason: FailedRunningDf,
+		Message: fmt.Sprintf(warningStrings, pvNames),
+	})
+}
+
+// getRefToAnalyticNs given a ns, returns the reference to correspoding ns present within MigAnalytic.Status.Analytics, stores found ones in cache
+func (pva *PersistentVolumeAdjuster) getRefToAnalyticNs(namespace string) *migapi.MigAnalyticNamespace {
+	if pva.StatusRefCache == nil {
+		pva.StatusRefCache = make(map[string]*migapi.MigAnalyticNamespace)
+	}
+	if nsRef, exists := pva.StatusRefCache[namespace]; exists {
+		return nsRef
+	}
+	for i := range pva.Owner.Status.Analytics.Namespaces {
+		nsRef := &pva.Owner.Status.Analytics.Namespaces[i]
+		if namespace == nsRef.Namespace {
+			pva.StatusRefCache[namespace] = nsRef
+			return nsRef
+		}
+	}
+	return nil
+}
+
+func (pva *PersistentVolumeAdjuster) calculateProposedVolumeSize(usagePercentage int64, actualCapacity resource.Quantity,
+	requestedCapacity resource.Quantity) (proposedSize resource.Quantity, reason string) {
+	return
```

Comment:
Assume this still needs to be implemented?

Reply:
@shawn-hurley correct. Shubham is implementing this piece. We will get this in a different PR